File: ios-native/Dimo/Auth/PKCE.swift
import Foundation
import CryptoKit

enum PKCE {
  static func makeVerifier() -> String {
    var bytes = [UInt8](repeating: 0, count: 32)
    _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
    return base64URLEncode(Data(bytes))
  }

  static func challenge(for verifier: String) -> String {
    let digest = SHA256.hash(data: Data(verifier.utf8))
    return base64URLEncode(Data(digest))
  }

  static func makeState() -> String {
    var bytes = [UInt8](repeating: 0, count: 16)
    _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
    return base64URLEncode(Data(bytes))
  }

  private static func base64URLEncode(_ data: Data) -> String {
    data.base64EncodedString()
      .replacingOccurrences(of: "+", with: "-")
      .replacingOccurrences(of: "/", with: "_")
      .replacingOccurrences(of: "=", with: "")
  }
}

File: android-native/app/src/main/java/app/dimo/android/auth/PKCE.kt
package app.dimo.android.auth

import android.util.Base64
import java.security.MessageDigest
import java.security.SecureRandom

/** Port of `ios-native/Dimo/Auth/PKCE.swift`. */
object PKCE {
  private val random = SecureRandom()

  fun makeVerifier(): String = base64UrlEncode(ByteArray(32).also { random.nextBytes(it) })

  fun challenge(verifier: String): String {
    val digest = MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray(Charsets.UTF_8))
    return base64UrlEncode(digest)
  }

  fun makeState(): String = base64UrlEncode(ByteArray(16).also { random.nextBytes(it) })

  private fun base64UrlEncode(bytes: ByteArray): String =
    Base64.encodeToString(bytes, Base64.NO_WRAP or Base64.NO_PADDING or Base64.URL_SAFE)
}
